docs/community/tutorials/bert-qa-benchmarking/index-bert-qa-benchmarking-with-redisai-and-redisgears.mdx

```diff
@@ -0,0 +1,406 @@
+
+Where I use two RedisGears cool features: capture event on key miss and async/await with the ability to run RedisAI on each shard and in non-locking for primary thread way - so Redis Cluster can continue to serve other customers. For benchmarks, caching RedisAI response is [disabled](https://github.com/applied-knowledge-systems/the-pattern-api/blob/156633b9934f1243775671ce6c18ff2bf471c0ce/qasearch/qa_redisai_keymiss_no_cache_np.py#L29) and if you are getting response time in nanoseconds on the second call rather then milliseconds you have deployed production version of RedisGears.
+
+# Running Benchmarks
+
+pre-requisite for running the benchmark:
+
+Assuming you are running Debian or ubuntu, have docker and docker-compose (or can create a virtual environment via conda):
+
+```
+git clone --recurse-submodules https://github.com/applied-knowledge-systems/the-pattern.git
+cd the-pattern
+./bootstrap_benchmark.sh
+```
+
+It should end with a curl call to qasearch API, Redis caching is disabled for the benchmark.
+
+Curl call shall look like this:
+
+```
+time curl -i -H "Content-Type: application/json" -X POST -d '{"search":"Who performs viral transmission among adults"}' http://localhost:8080/qasearch
+```
+
+with output:
+
+```
+HTTP/1.1 200 OK
+Server: nginx/1.18.0 (Ubuntu)
+Date: Sun, 29 May 2022 12:05:39 GMT
+Content-Type: application/json
+Content-Length: 2120
+Connection: keep-alive
+
+{"links":[{"created_at":"2002","rank":13,"source":"C0001486","target":"C0152083"}],"results":[{"answer":"adenovirus","sentence":"The medium of 40 T150 flasks of adenovirus transducer dec CAR CHO cells yielded 0 5 1 my of purified msCEACAM1a 1 4 protein","sentencekey":"sentence:PMC125375.xml:{mG}:202","title":"Crystal structure of murine sCEACAM1a[1,4]: a coronavirus receptor in the CEA family"}] OUTPUT_REDUCTED}
+```
+
+I modified the output of API for the benchmark to return results from all shards - even if the answer is empty, in the run above five shards return answers, overall API call response under second with all additional hops to search in RedisGraph.
+
+![Architecture Diagram for BERT QA API call](diagrams_01.png)
+
+# Deep Dive into Benchmark
+
+Let's dig deeper into what's happening under the hood:
+
+There is a sentence key with shard id or grab "Cache key" from `docker logs -f rgcluster`, in my setup cache key "bertqa{6fd}\_PMC169038.xml:{6fd}:33_Who performs viral transmission among adults". If you think it looks like a function call - it's a function call, which is triggered if key isn't present in Redis Cluster, which is for benchmark will be every time, since we disabled saving output for cache in a previous paragraph.
+
+One more thing is to figure out from logs the port of the shard corresponding to hashtag (also known as shard id, stuff in curly brackets – like this {6fd}, same will be in the output for export_load script, in my case Cache key was in "30012.log" so my port 30012
```

Comment:
```suggestion
One more thing is to figure out from the logs the port of the shard corresponding to the hashtag (also known as shard ID: the part of the key name in curly brackets, for example `{6fd}`.  This will be in the output for the export_load script, in my case the Cache key was in "30012.log" so my port is 30012.
```

This is confusing to follow even after editing.